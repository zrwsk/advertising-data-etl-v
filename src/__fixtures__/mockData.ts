const mockData = [
  {
    date: '01.01.2020',
    campaign: 'C1',
    datasource: 'D1',
    clicks: 0,
    impressions: 232,
  },
  {
    date: '02.01.2020',
    campaign: 'C2',
    datasource: 'D3',
    clicks: 1000,
    impressions: 2112,
  },
  {
    date: '01.01.2020',
    campaign: 'C1',
    datasource: 'D3',
    clicks: 543592,
    impressions: 2342,
  },
  {
    date: '01.02.2020',
    campaign: 'C2',
    datasource: 'D1',
    clicks: 123,
    impressions: 3422,
  },
  {
    date: '03.03.2020',
    campaign: 'C4',
    datasource: 'D3',
    clicks: 4342,
    impressions: 4432,
  },
  {
    date: '04.05.2020',
    campaign: 'C3',
    datasource: 'D6',
    clicks: 5435435,
    impressions: 24324,
  },
  {
    date: '03.03.2020',
    campaign: 'C1',
    datasource: 'D2',
    clicks: 0,
    impressions: 1232,
  },
];

export default mockData;
