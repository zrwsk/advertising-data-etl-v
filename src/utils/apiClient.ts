import { parse } from 'papaparse';
import { IAdDataEntity } from '../types';

const URL = 'http://adverity-challenge.s3-website-eu-west-1.amazonaws.com/DAMKBAoDBwoDBAkOBAYFCw.csv';

export const getAdvertisingData = (): Promise<IAdDataEntity[]> => new Promise(
  (resolve, reject) => parse<IAdDataEntity>(URL, {
    header: true,
    download: true,
    skipEmptyLines: true,
    transformHeader: (header) => header.toLowerCase(),
    // Papa parser has dynamic typing option, which works for the most of cases
    // but returns null for empty values. In our case we want explicitly
    // to have clicks and impressions as a number and in case of empty value it should be zero.
    transform: (value, field) => (['clicks', 'impressions'].includes(field as string) ? Number(value) : value),
    error: (err) => reject(err.message),
    complete: (result) => resolve(result.data),
  }),
);
