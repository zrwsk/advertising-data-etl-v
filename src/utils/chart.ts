import { IFilters } from '../types';

const getFilterOptionsText = (options: (string | number)[]) => options.map((o) => `"${o}"`).join(', ').replace(/, ([^,]*)$/, ' and $1');

export const getHeadingText = ({ datasources, campaigns }: IFilters): string => {
  const datasourcesText = datasources.length > 0 ? `Datasource ${getFilterOptionsText(datasources)}` : 'All Datasources';
  const campaignsText = campaigns.length > 0 ? `Campaign ${getFilterOptionsText(campaigns)}` : 'All Campaigns';

  return `${datasourcesText}; ${campaignsText}`;
};

export const formatYAxisTick = (v: unknown): string => (typeof v === 'number' && v > 10000 ? `${Math.round(v / 1000)}k` : String(v));
