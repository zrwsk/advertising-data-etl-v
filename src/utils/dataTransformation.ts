import { chain, sumBy } from 'lodash';
import { IAdDataEntity, IFilters } from '../types';

export const getFilterOptions = (
  data: IAdDataEntity[],
  key: keyof IAdDataEntity,
): (string | number)[] => chain(data)
  .map((dataPoint) => dataPoint[key])
  .uniq()
  .value();

const includesOrEmpty = <T>(arr: T[], item: T) => (arr.length > 0 ? arr.includes(item) : true);

export const getAggregatedData = (
  data: IAdDataEntity[],
  filters: IFilters,
): Pick<IAdDataEntity, 'date' | 'clicks' | 'impressions'>[] => chain(data)
  .filter(({ datasource, campaign }) => includesOrEmpty(filters.datasources, datasource)
    && includesOrEmpty(filters.campaigns, campaign))
  .groupBy((dataPoint) => dataPoint.date)
  .map((group, key) => ({
    date: key,
    clicks: sumBy(group, 'clicks'),
    impressions: sumBy(group, 'impressions'),
  }))
  .sortBy((dataPoint) => dataPoint.date?.split('.').reverse().join(''))
  .value();
