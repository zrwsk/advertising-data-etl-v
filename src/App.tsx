import { useEffect, useState } from 'react';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Filters from './components/Filters';
import Chart from './components/Chart';
import { getAdvertisingData } from './utils/apiClient';
import { LoadingStates } from './enums';
import { IAdDataEntity } from './types';
import LoadingManager from './components/LoadingManager';

const App = (): JSX.Element => {
  const [data, setData] = useState<IAdDataEntity[]>([]);
  const [loadingState, setLoadingState] = useState<LoadingStates>(LoadingStates.IDLE);

  useEffect(() => {
    setLoadingState(LoadingStates.FETCHING);

    getAdvertisingData()
      .then((v) => {
        setData(v);
        setLoadingState(LoadingStates.FULFILLED);
      })
      .catch(() => {
        setLoadingState(LoadingStates.ERROR);
      });
  }, []);

  return (
    <>
      <LoadingManager loadingState={loadingState} />
      <Box m={4}>
        <Typography variant="h3" component="h1" gutterBottom>Advertising Data ETL-V</Typography>
        <Filters />
        <Chart />
      </Box>
    </>
  );
};

export default App;
