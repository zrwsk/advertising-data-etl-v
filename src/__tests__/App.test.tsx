import {
  render, screen, waitForElementToBeRemoved, fireEvent,
} from '@testing-library/react';
import App from '../App';
import * as apiClient from '../utils/apiClient';
import mockData from '../__fixtures__/mockData';

describe('App component', () => {
  beforeEach(() => {
    jest.spyOn(apiClient, 'getAdvertisingData').mockResolvedValue(mockData);
  });

  it('should show loader when data is loading', async () => {
    render(<App />);

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
    // after rerender progressbar should disappear
    expect(await screen.findByRole('progressbar')).not.toBeInTheDocument();
  });

  it('should show error if data can not be loaded', async () => {
    (apiClient.getAdvertisingData as jest.Mock).mockRejectedValueOnce('message');
    render(<App />);

    expect(await screen.findByText('Error while fetching the data')).toBeInTheDocument();
  });

  it('should show all the data', async () => {
    render(<App />);

    await waitForElementToBeRemoved(screen.getByRole('progressbar'));
    expect(screen.getByTestId('chart-data')).toMatchSnapshot();
    expect(screen.getByText('All Datasources; All Campaigns')).toBeInTheDocument();
  });

  it('should show filtered data', async () => {
    render(<App />);

    await waitForElementToBeRemoved(screen.getByRole('progressbar'));

    fireEvent.click(screen.getAllByTitle('Open')[0]);
    fireEvent.click(screen.getByText('D1'));
    fireEvent.click(screen.getAllByTitle('Open')[1]);
    fireEvent.click(screen.getByText('C2'));
    fireEvent.click(screen.getByText('Apply'));

    expect(screen.getByTestId('chart-data')).toMatchSnapshot();
    expect(screen.getByText('Datasource "D1"; Campaign "C2"')).toBeInTheDocument();
  });
});
