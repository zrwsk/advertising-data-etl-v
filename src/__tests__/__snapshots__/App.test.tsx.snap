// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should show all the data 1`] = `
<ul
  data-testid="chart-data"
>
  <li>
    01.01.2020, 543592, 2574
  </li>
  <li>
    02.01.2020, 1000, 2112
  </li>
  <li>
    01.02.2020, 123, 3422
  </li>
  <li>
    03.03.2020, 4342, 5664
  </li>
  <li>
    04.05.2020, 5435435, 24324
  </li>
</ul>
`;

exports[`App component should show filtered data 1`] = `
<ul
  data-testid="chart-data"
>
  <li>
    01.02.2020, 123, 3422
  </li>
</ul>
`;
